src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -337,49 +286,213 @@ public virtual string CreateToken(string payload, SigningCredentials signingCred
+            string tokenType,
+            IDictionary<string, object> jwsHeaderClaims,
+            IDictionary<string, object> jweHeaderClaims)
+        {
+            using (MemoryStream memoryStream = new MemoryStream())
+            {
+                Utf8JsonWriter writer = null;
+                try
+                {
+                    writer = new Utf8JsonWriter(memoryStream, new JsonWriterOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
+                    writer.WriteStartObject();
+
+                    if (signingCredentials == null)
+                    {
+                        writer.WriteString(JwtHeaderUtf8Bytes.Alg, SecurityAlgorithms.None);
+                    }
+                    else
+                    {
+                        writer.WriteString(JwtHeaderUtf8Bytes.Alg, signingCredentials.Algorithm);
+                        if (signingCredentials.Key.KeyId != null)
+                            writer.WriteString(JwtHeaderUtf8Bytes.Kid, signingCredentials.Key.KeyId);
+
+                        if (signingCredentials.Key is X509SecurityKey x509SecurityKey)
+                            writer.WriteString(JwtHeaderUtf8Bytes.X5t, x509SecurityKey.X5t);
+                    }
+
+                    bool useJwsHeaderClaims = jwsHeaderClaims != null && jwsHeaderClaims.Count > 0;
+                    bool typeWritten = false;
+
+                    // Priority is jwsHeaderClaims, jweHeaderClaims, default
+                    if (jweHeaderClaims != null && jweHeaderClaims.Count > 0)
+                    {
+                        foreach (KeyValuePair<string, object> kvp in jweHeaderClaims)
+                        {
+                            if (useJwsHeaderClaims && jwsHeaderClaims.ContainsKey(kvp.Key))
+                                continue;
+
+                            JsonPrimitives.WriteObject(ref writer, kvp.Key, kvp.Value);
+                            if (!typeWritten && kvp.Key.Equals(JwtHeaderParameterNames.Typ, StringComparison.Ordinal))
+                                typeWritten = true;
+                        }
+                    }
+
+                    if (useJwsHeaderClaims)
+                    {
+                        foreach (KeyValuePair<string, object> kvp in jwsHeaderClaims)
+                        {
+                            JsonPrimitives.WriteObject(ref writer, kvp.Key, kvp.Value);
+                            if (!typeWritten && kvp.Key.Equals(JwtHeaderParameterNames.Typ, StringComparison.Ordinal))
+                                typeWritten = true;
+                        }
+                    }
+
+                    if (!typeWritten)
+                        writer.WriteString(JwtHeaderUtf8Bytes.Typ, string.IsNullOrEmpty(tokenType) ? JwtConstants.HeaderType : tokenType);
+
+                    writer.WriteEndObject();
+                    writer.Flush();
+
+                    return memoryStream.ToArray();
```

Comment:
here and below, can we use the memory streams buffer instead of allocating a new array?

Reply:
You could do that if you pass in the MemoryBuffer to this function. Since there is only 1 call site and it immediately uses and throws away the buffer after getting a string that's a nice easy allocation saving.